File: Algorithms/基础/链表/LC707.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"
#define lowbit(x) (x & (-x))
typedef long long LL;
typedef unsigned long long ULL;
typedef pair<int, int> PII;
const int INF = 0x3f3f3f3f, MOD = 1e9 + 7;

// https://leetcode.cn/problems/design-linked-list/

class MyLinkedList {
    struct Node {
        int val;
        Node* next;
        Node(int val) : val(val), next(nullptr) {}
        Node(int val, Node* next) : val(val), next(next) {}
    };
    int length;
    Node *head, *tail;

public:
    MyLinkedList() : length(0), head(nullptr), tail(nullptr) {}

    int get(int index) {
        if (index < 0 || index >= length) return -1;
        auto p = head;
        while (index--) p = p->next;
        return p->val;
    }

    void addAtHead(int val) {
        if (length)
            head = new Node(val, head);
        else
            head = tail = new Node(val);
        length++;
    }

    void addAtTail(int val) {
        if (length) {
            tail->next = new Node(val);
            tail = tail->next;
        } else
            head = tail = new Node(val);
        length++;
    }

    void addAtIndex(int index, int val) {
        if (index > length)
            return;
        else if (index <= 0)
            addAtHead(val);
        else if (index == length)
            addAtTail(val);
        else {
            auto p = head;
            while (--index) p = p->next;
            p->next = new Node(val, p->next);
            length++;
        }
    }

    void deleteAtIndex(int index) {
        if (index < 0 || index >= length)
            return;
        else if (length == 1)
            head = tail = nullptr;
        else if (index == 0)
            head = head->next;
        else {
            auto p = head;
            for (int i = 0; i < index - 1; ++i) p = p->next;
            p->next = p->next->next;
            if (index == length - 1) tail = p;
        }
        length--;
    }
};

int main() {
    

    return 0;
}

File: 数据结构/基础/链表/lc707.go
package main

// https://leetcode.cn/problems/design-linked-list/description/

type node struct {
	prev, next *node
	val        int
}

type MyLinkedList struct {
	dummy *node
	n     int
}

func Constructor() MyLinkedList {
	dummy := &node{}
	dummy.prev = dummy
	dummy.next = dummy
	return MyLinkedList{dummy: dummy}
}

func (this *MyLinkedList) Get(index int) int {
	if index >= this.n {
		return -1
	}
	p := this.dummy.next
	for ; index > 0; index-- {
		p = p.next
	}
	return p.val
}

func (this *MyLinkedList) add(p *node, val int) {
	x := &node{val: val}
	x.next = p.next
	x.prev = p
	x.next.prev = x
	x.prev.next = x
	this.n++
}

func (this *MyLinkedList) AddAtHead(val int) { this.add(this.dummy, val) }

func (this *MyLinkedList) AddAtTail(val int) { this.add(this.dummy.prev, val) }

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index > this.n {
		return
	}
	if index == this.n {
		this.AddAtTail(val)
		return
	}
	p := this.dummy
	for ; index > 0; index-- {
		p = p.next
	}
	this.add(p, val)
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index >= this.n {
		return
	}
	p := this.dummy.next
	for ; index > 0; index-- {
		p = p.next
	}
	p.prev.next = p.next
	p.next.prev = p.prev
	p.next = nil
	p.prev = nil
	this.n--
}
